xgi/algorithms/centrality.py

```diff
@@ -392,3 +387,161 @@ def katz_centrality(H, cutoff=100):
+    if H.num_nodes == 0:
+        return dict()
+    # if the hypergraph is not connected,
+    # this metric doesn't make sense and should return nan.
+    if not is_connected(H):
+        return {n: np.nan for n in H.nodes}
+
+    new_H = convert_labels_to_integers(H, "old-label")
+    edge_dict = new_H.edges.members(dtype=dict)
+    node_dict = new_H.nodes.memberships()
+    r = new_H.edges.size.max()
+
+    x = np.random.uniform(size=(new_H.num_nodes))
+    x = x / norm(x, 1)
+    y = np.abs(np.array(ttsv1(node_dict, edge_dict, r, x)))
+
+    converged = False
+    it = 0
+    while it < max_iter and not converged:
+        y_scaled = [_y ** (1 / (r - 1)) for _y in y]
+        x = y_scaled / norm(y_scaled, 1)
+        y = np.abs(np.array(ttsv1(node_dict, edge_dict, r, x)))
+        s = [a / (b ** (r - 1)) for a, b in zip(y, x)]
+        converged = (np.max(s) - np.min(s)) / np.min(s) < tol
+        if converged:
+            break
+        it += 1
+    else:
+        warn("Iteration did not converge!")
+    return {
+        new_H.nodes[n]["old-label"]: c.item()
+        for n, c in zip(new_H.nodes, x / norm(x, 1))
+    }
+
+
+def z_eigenvector_tensor_centrality(H, max_iter=100, tol=1e-6):
+    """Compute the Z-eigenvector centrality of a hypergraph.
+
+    Parameters
+    ----------
+    H : Hypergraph
+        The hypergraph of interest.
+    max_iter : int, optional
+        The maximum number of iterations before the algorithm terminates.
+        By default, 100.
+    tol : float > 0, optional
+        The desired convergence tolerance. By default, 1e-6.
+
+    Returns
+    -------
+    dict
+        Centrality, where keys are node IDs and values are centralities. The
+        centralities are 1-normalized.
+
+    Raises
+    ------
+    XGIError
+        If the hypergraph is not uniform.
+
+    See Also
+    --------
+    clique_eigenvector_centrality
+
+    References
+    ----------
+    Scalable Tensor Methods for Nonuniform Hypergraphs,
+    Sinan Aksoy, Ilya Amburg, Stephen Young,
+    https://doi.org/10.1137/23M1584472
+
+    Three Hypergraph Eigenvector Centralities,
+    Austin R. Benson,
+    https://doi.org/10.1137/18M1203031
+    """
+    # if there aren't any nodes, return an empty dict
+    n = H.num_nodes
+    if n == 0:
+        return dict()
+
+    # if the hypergraph is not connected,
+    # this metric doesn't make sense and should return nan.
+    if not is_connected(H):
+        return {n: np.nan for n in H.nodes}
+    new_H = convert_labels_to_integers(H, "old-label")
+    edge_dict = new_H.edges.members(dtype=dict)
+    pairs_dict = pairwise_incidence(edge_dict)
+
+    r = H.edges.size.max()
+
+    def LR_evec(A):
+        """Compute the largest real eigenvalue of the matrix A"""
+        _, v = eigsh(A, k=1, which="LM", tol=1e-5, maxiter=200)
+        evec = np.array([_v for _v in v[:, 0]])
+        if evec[0] < 0:
+            evec = -evec
+        return evec / norm(evec, 1)
+
+    def f(u):
```

Comment:
what does this do?

Reply:
This makes the eigenvector positive if you happen to compute a negative one and makes it stochastic.